tiledb/sm/metadata/metadata.cc

```diff
@@ -0,0 +1,279 @@
+/**
+ * @file   metadata.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2019 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class Metadata.
+ */
+
+#include "tiledb/sm/metadata/metadata.h"
+#include "tiledb/sm/misc/logger.h"
+#include "tiledb/sm/misc/utils.h"
+
+#include <iostream>
+
+namespace tiledb {
+namespace sm {
+
+/* ********************************* */
+/*     CONSTRUCTORS & DESTRUCTORS    */
+/* ********************************* */
+
+Metadata::Metadata() {
+  auto t = utils::time::timestamp_now_ms();
+  timestamp_range_ = std::make_pair(t, t);
+}
+
+Metadata::~Metadata() = default;
+
+/* ********************************* */
+/*                API                */
+/* ********************************* */
+
+void Metadata::clear() {
+  metadata_map_.clear();
+  metadata_index_.clear();
+  loaded_metadata_uris_.clear();
+  timestamp_range_ = std::make_pair(0, 0);
+}
+
+Status Metadata::deserialize(
+    const std::vector<std::shared_ptr<ConstBuffer>>& metadata_buffs) {
+  clear();
+
+  if (metadata_buffs.empty())
+    return Status::Ok();
+
+  uint32_t key_len;
+  std::string key;
+  char del;
+  char value_type;
+  uint32_t value_num;
+  size_t value_len, value_vec_len, offset;
+  std::vector<uint8_t> value, value_vec;
+  for (const auto& buff : metadata_buffs) {
+    // Iterate over all items
+    buff->set_offset(0);
+    while (buff->offset() != buff->size()) {
+      RETURN_NOT_OK(buff->read(&key_len, sizeof(uint32_t)));
+      key.resize(key_len);
+      RETURN_NOT_OK(buff->read((void*)key.data(), key_len));
+      RETURN_NOT_OK(buff->read(&del, sizeof(char)));
+
+      // Handle deletion
+      if (del) {
+        metadata_map_.erase(key);
+        continue;
```

Comment:
Which loop should this `continue` affect?

Reply:
Always the inner, i.e., it will advance the iteration to the next item in the current buffer (which corresponds to a separate metadata file, similar to array fragments).

Reply:
Even if it's a deletion don't you still need to skip past the rest of the entry in the buffer before you can do `continue` here?

Reply:
A deletion entry does not have anything else after `del` :). I am planning to add the format spec to the docs in the next commit.